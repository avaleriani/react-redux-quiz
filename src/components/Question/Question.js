import React, { Component } from "react";
import Btn from "../Btn/Btn";
import { cleanHtmlText } from "../../Utils/utils";
import styles from "./styles.module.css";

/**
 * @param {{question: Object, onSubmit: function, userAnswerState: function}} props
 * @return HTMLElement
 */
class Question extends Component {
  constructor() {
    super();
    this.state = {
      answer: '',
      isWrong: false
    };

    this.onTextUpdate = this.onTextUpdate.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  onTextUpdate(event) {
    this.setState({
      answer: event.target.value
    });
  }

  handleSubmit(event) {
    event.preventDefault();
    if (this.state.answer) {
      //Sometimes the answer come wrapped in HTML.
      const correctAnswer = cleanHtmlText(this.props.question.answer);
      if (this.isAnswerCorrect(correctAnswer, this.state.answer)) {
        this.setState({
          isWrong: false
        });
        this.props.onSubmit(this.props.question.answer, this.state.answer)
      } else {
        this.setState({
          isWrong: true
        });
        this.props.userAnswerState(false);
      }
    }
  };

  /**
   * @type {string} correctAnswer
   * @type {string} answer
   * @return HTMLElement
   */
  isAnswerCorrect(correctAnswer, answer) {
    return correctAnswer.toLowerCase() === answer.toLowerCase();
  }

  render() {
    return this.props.question ? (
      <form onSubmit={this.handleSubmit}>
        <div className={styles['question']}>
          <div className={styles['question-category']}>
            {this.props.question ? this.props.question.category.title : ''}
          </div>
          <div className={styles['question-text']}>
            {this.props.question.question}
          </div>
        </div>
        <input placeholder="Your answer" className={styles['answer']} type="text" onChange={this.onTextUpdate}
               autoFocus/>
        <div className={styles['button-container']}>
          {this.state.isWrong ? <div className={styles['is-wrong']}>
            <span role="img" aria-label="thumbs-down">ðŸ‘Ž</span>
            Wrong answer
            <span role="img" aria-label="thumbs-down"> ðŸ‘Ž</span>
          </div> : null}
          <Btn type="submit" text="Submit"/>
        </div>
      </form>
    ) : null
  }
}

export default Question;