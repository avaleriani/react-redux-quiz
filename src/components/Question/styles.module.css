.question {
    border-top: 2px solid #69B578;
    border-left: 2px solid #69B578;
    border-right: 2px solid #69B578;
    margin: 30px 0 0 0;
    padding: 25px;
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
    min-width: 300px;
    height: 175px;
}

.question-category {
    padding-bottom: 40px;
    text-transform: capitalize;
    font-size: 1.25rem;
}

.question-text {
    padding-bottom: 40px;
}

.answer {
    width: calc(100% - 24px);
    height: 35px;
    line-height: 35px;
    outline: none;
    font-size: 1.25rem;
    border: 2px solid #69B578;
    margin-bottom: 35px;
    color: #ffffff;
    padding: 0 10px 0 10px;
    background-color: rgba(40, 44, 52, 0.8);
}

.button-container {
    display: flex;
    min-height: 90px;
    flex-direction: column;
}

.is-wrong {
    color: #C92828;
    text-align: center;
    opacity: 0;
    animation: fadein 760ms forwards;
}

@keyframes fadein {
    from {
        opacity: 0;
    }
    to {
        opacity: 1;
    }
}