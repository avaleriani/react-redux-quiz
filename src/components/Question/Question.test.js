import React from 'react';
import renderer from 'react-test-renderer';
import Question from './Question';

describe('Question component', () => {

  test('renders correctly without props', () => {
    const QuestionComponent = renderer.create(<Question/>).toJSON();
    expect(QuestionComponent).toMatchSnapshot();
  });

  test('renders with props', () => {
    const question = {
      "id": 5129,
      "answer": "SALT II",
      "question": "Acronym used to refer to the treaty signed by Brezhnev \u0026 Carter in Vienna in June",
      "value": 100,
      "airdate": "1988-11-10T12:00:00.000Z",
      "created_at": "2014-02-11T22:49:55.954Z",
      "updated_at": "2014-02-11T22:49:55.954Z",
      "category_id": 711,
      "game_id": null,
      "invalid_count": null,
      "category": {
        "id": 711,
        "title": "1979",
        "created_at": "2014-02-11T22:49:55.844Z",
        "updated_at": "2014-02-11T22:49:55.844Z",
        "clues_count": 10
      }
    };
    const props = {
      question: question,
      onSubmit: () => {
      },
      userAnswerState: () => {
      }
    };
    const QuestionComponent = renderer.create(<Question {...props}/>).toJSON();
    expect(QuestionComponent).toMatchSnapshot();
  });
});