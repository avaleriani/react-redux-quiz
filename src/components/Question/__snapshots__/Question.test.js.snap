// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Question component renders correctly without props 1`] = `null`;

exports[`Question component renders with props 1`] = `
<form
  onSubmit={[Function]}
>
  <div
    className="question"
  >
    <div
      className="question-category"
    >
      1979
    </div>
    <div
      className="question-text"
    >
      Acronym used to refer to the treaty signed by Brezhnev & Carter in Vienna in June
    </div>
  </div>
  <input
    autoFocus={true}
    className="answer"
    onChange={[Function]}
    placeholder="Your answer"
    type="text"
  />
  <div
    className="button-container"
  >
    <button
      className="button"
      onClick={null}
      type="submit"
    >
      Submit
    </button>
  </div>
</form>
`;
